### PR TITLE
Add unchanged names to existence filter and watchFilters spec test runner

### PR DESCRIPTION
**Pull request by milaGGL**
#no-changelog

Ported from Web SDK: https://github.com/firebase/firebase-js-sdk/pull/6839
- [x] mark `unchangedNames` and `bits` optional in proto
- [x] Add `bloom_filter` to `ExistenceFilter` class
- [x] Update serializer to pass `unchangedNames` to `ExistenceFilter`
- [x] update the Spec test  runner `doWatchFilter` to accept `NSDictionary` instead of `NSArray`
  - In web SDK spec test runner `doWatchFilter(watchFilter: SpecWatchFilter)`, the specWatchFilter input type changed from an array to object:
Screenshot 2023-02-23 at 5 46 34 PM
- [x] Ported the spec tests that are changed due to the `doWatchFilter` change above(no bloom filter proto json values are included)

### CODE REVIEW

**Review by dconeybe — CHANGES_REQUESTED** [review]
We need to fix the proto passing and memory management issue in this PR. I haven't reviewed all of this PR but wanted to send the feedback now so you can have a chance to review prior to our meeting in an hour. The logic you want to achieve in this PR is a bit more difficult than it had appeared on the surface!

File: Firestore/Example/Tests/SpecTests/FSTSpecTests.mm [hunk 1]
```diff
@@ -325,6 +330,30 @@ - (SnapshotVersion)parseVersion:(NSNumber *_Nullable)version {
...
+  NSString *bitmapEncoded = bitsData[@"bitmap"];
+  std::string bitmapDecoded;
+  absl::Base64Unescape([bitmapEncoded UTF8String], &bitmapDecoded);
+  std::vector<uint8_t> bitmap(bitmapDecoded.begin(), bitmapDecoded.end());
+
+  // If not specified, will default padding and hashCount to 0.
+  int32_t padding = [bitsData[@"padding"] intValue];
+  int32_t hashCount = [bloomFilterProto[@"hashCount"] intValue];
+  StatusOr<BloomFilter> maybeBloomFilter = BloomFilter::Create(bitmap, padding, hashCount);
+  if (maybeBloomFilter.ok()) {
```
Comment (milaGGL) [inline Firestore/Example/Tests/SpecTests/FSTSpecTests.mm:350]: TODO: log the error message if the bloomFilter json object has invalid inputs.

File: Firestore/Protos/protos/google/firestore/v1/write.options [hunk 2]
```diff
@@ -29,3 +29,7 @@ google.firestore.v1.Write.update_mask proto3:false
 
 # update_time should not be set for deletes.
 google.firestore.v1.WriteResult.update_time proto3:false
+
+# `unchanged_names` field might not be set if ListenRequest doesn't have
+# `expected_count` field.
+google.firestore.v1.ExistenceFilter.unchanged_names proto3:false
```
Comment (milaGGL) [inline Firestore/Protos/protos/google/firestore/v1/write.options:35]: `unchanged_names` might not be included in the listen response, and we need to check its existence to create a BloomFilter or leave it null.

Or, we can give default values to the missing  `unchanged_names`, and end up with an empty BloomFilter.

File: Firestore/core/src/remote/serializer.cc [hunk 3]
```diff
@@ -1425,9 +1426,31 @@ std::unique_ptr<WatchChange> Serializer::DecodeDocumentRemove(
...
+  return absl::make_unique<ExistenceFilterWatchChange>(
+      std::move(existence_filter), filter.target_id);
+}
+
+ExistenceFilter Serializer::DecodeExistenceFilter(
+    const google_firestore_v1_ExistenceFilter& filter) const {
+  // Create bloom filter if there is an unchanged_names present in the filter
+  // and inputs are valid, otherwise keep it null.
+  absl::optional<BloomFilter> bloom_filter = absl::nullopt;
+  if (filter.has_unchanged_names && filter.unchanged_names.has_bits) {
```
Comment (milaGGL) [inline Firestore/core/src/remote/serializer.cc:1439]: This is where we check the existence of the `unchanged_names` and 'bits` that we made optional in proto.

File: Firestore/core/src/remote/watch_change.h [hunk 4]
```diff
@@ -115,7 +115,7 @@ bool operator==(const DocumentWatchChange& lhs, const DocumentWatchChange& rhs);
 class ExistenceFilterWatchChange : public WatchChange {
  public:
   ExistenceFilterWatchChange(ExistenceFilter filter, model::TargetId target_id)
-      : filter_{filter}, target_id_{target_id} {
+      : filter_{std::move(filter)}, target_id_{target_id} {
```
Comment (milaGGL) [inline Firestore/core/src/remote/watch_change.h:118]: did not move the ExistenceFilter before cause it only includes `count`, now it might include `unchanged_names`

File: Firestore/core/src/remote/bloom_filter.h [hunk 5]
```diff
@@ -87,23 +104,15 @@ class BloomFilter final {
...
-   * The number of hash functions used to construct the filter. Guaranteed to
-   * be non-negative.
-   */
   int32_t hash_count_ = 0;
 
-  /** Bloom filter's bitmap. */
   std::vector<uint8_t> bitmap_;
 };
 
+bool operator==(const BloomFilter& lhs, const BloomFilter& rhs);
```
Comment (dconeybe) [inline Firestore/core/src/remote/bloom_filter.h:114]: Add `operator!=()` as well, for symmetry. It's very simple, for example: https://github.com/firebase/firebase-ios-sdk/blob/b2383a9d3ca37b6a9e99c5fd9b75abdcf5a497c8/Firestore/core/src/model/transform_operation.h#L226-L229

File: Firestore/core/src/remote/serializer.cc [hunk 6]
```diff
@@ -1429,9 +1430,29 @@ std::unique_ptr<WatchChange> Serializer::DecodeDocumentRemove(
...
+    // actual parsing code will be written in the next PR, where we can validate
+    // the parsing result.
+    StatusOr<BloomFilter> maybe_bloom_filter =
+        BloomFilter::Create(std::vector<uint8_t>{}, 0, 0);
+    if (maybe_bloom_filter.ok()) {
+      bloom_filter = maybe_bloom_filter.ValueOrDie();
+    }
+  }
+
+  return {filter.count, bloom_filter};
```
Comment (dconeybe) [inline Firestore/core/src/remote/serializer.cc:1455]: nit: `bloom_filter` -> `std::move(bloom_filter)`

File: Firestore/core/src/remote/serializer.cc [hunk 7]
```diff
@@ -1429,9 +1430,29 @@ std::unique_ptr<WatchChange> Serializer::DecodeDocumentRemove(
...
+  ExistenceFilter existence_filter = DecodeExistenceFilter(filter);
+  return absl::make_unique<ExistenceFilterWatchChange>(
+      std::move(existence_filter), filter.target_id);
+}
+
+ExistenceFilter Serializer::DecodeExistenceFilter(
+    const google_firestore_v1_ExistenceFilter& filter) const {
+  // Create bloom filter if there is an unchanged_names present in the filter
+  // and inputs are valid, otherwise keep it null.
+  absl::optional<BloomFilter> bloom_filter = absl::nullopt;
```
Comment (dconeybe) [inline Firestore/core/src/remote/serializer.cc:1442]: Is it possible to omit `= absl::nullopt` in this line? I think it ma be redundant, because I think that a default-constructed `absl::optional` object is "empty".

File: Firestore/core/src/remote/serializer.cc [hunk 8]
```diff
@@ -1429,9 +1430,29 @@ std::unique_ptr<WatchChange> Serializer::DecodeDocumentRemove(
...
+  if (filter.has_unchanged_names && filter.unchanged_names.has_bits) {
+    // TODO(Mila): None of the ported spec tests here actually has the bloom
+    // filter json string, so hard code an empty bloom filter for now. The
+    // actual parsing code will be written in the next PR, where we can validate
+    // the parsing result.
+    StatusOr<BloomFilter> maybe_bloom_filter =
+        BloomFilter::Create(std::vector<uint8_t>{}, 0, 0);
+    if (maybe_bloom_filter.ok()) {
+      bloom_filter = maybe_bloom_filter.ValueOrDie();
```
Comment (dconeybe) [inline Firestore/core/src/remote/serializer.cc:1451]: nit: `maybe_bloom_filter` -> `std::move(maybe_bloom_filter)`.

https://github.com/firebase/firebase-ios-sdk/blob/b2383a9d3ca37b6a9e99c5fd9b75abdcf5a497c8/Firestore/core/src/util/statusor.h#L189-L192

File: Firestore/Example/Tests/SpecTests/FSTSpecTests.mm [hunk 9]
```diff
@@ -463,13 +476,16 @@ - (void)doWatchEntity:(NSDictionary *)watchEntity {
...
   HARD_ASSERT(targets.count == 1, "ExistenceFilters currently support exactly one target only.");
 
-  int keyCount = watchFilter.count == 0 ? 0 : (int)watchFilter.count - 1;
+  NSArray<NSNumber *> *keys = watchFilter[@"keys"];
+  int keyCount = keys ? (int)keys.count : 0;
+
+  absl::optional<BloomFilter> bloomFilter = [self parseBloomFilter:watchFilter[@"bloomFilter"]];
 
-  ExistenceFilter filter{keyCount};
+  ExistenceFilter filter{keyCount, bloomFilter};
```
Comment (dconeybe) [inline Firestore/Example/Tests/SpecTests/FSTSpecTests.mm:488]: nit: `bloomFilter` -> `std::move(bloomFilter)`

File: Firestore/Example/Tests/SpecTests/FSTSpecTests.mm [hunk 10]
```diff
@@ -463,13 +476,16 @@ - (void)doWatchEntity:(NSDictionary *)watchEntity {
...
   ExistenceFilterWatchChange change{filter, targets[0].intValue};
```
Comment (dconeybe) [inline Firestore/Example/Tests/SpecTests/FSTSpecTests.mm:489]: nit: `filter` -> `std::move(filter)`

File: Firestore/core/test/unit/remote/bloom_filter_test.cc [hunk 11]
```diff
@@ -118,7 +118,20 @@ TEST(BloomFilterUnitTest, CreateShouldReturnNotOKStatusIfPaddingIsTooLarge) {
   EXPECT_EQ(maybe_bloom_filter.status().error_message(), "Invalid padding: 8");
 }
 
-TEST(BloomFilterUnitTest, MightContainCanProcessNonStandardCharacters) {
+TEST(BloomFilterTest, CanCheckBloomFiltersEquality) {
```
Comment (dconeybe) [inline Firestore/core/test/unit/remote/bloom_filter_test.cc:121]: nit: Add a test for `operator!=` as well. This test doesn't need to verify the _entire_ domain of possibilities, like the tests for operator== does. Just do one check when operator!= returns true, and another when it returns false.

File: Firestore/core/test/unit/remote/remote_event_test.cc
Comment (dconeybe) [inline Firestore/core/test/unit/remote/remote_event_test.cc:1]: Please add "TODO" comments in remote_event_test.cc, serializer_test.cc, and watch_change_test.cc to add test coverage for when the bloom filter is not null (if such tests do indeed make sense).

File: Firestore/core/src/remote/serializer.cc [hunk 12]
```diff
@@ -1429,9 +1430,29 @@ std::unique_ptr<WatchChange> Serializer::DecodeDocumentRemove(
...
+  // and inputs are valid, otherwise keep it null.
+  absl::optional<BloomFilter> bloom_filter = absl::nullopt;
+  if (filter.has_unchanged_names && filter.unchanged_names.has_bits) {
+    // TODO(Mila): None of the ported spec tests here actually has the bloom
+    // filter json string, so hard code an empty bloom filter for now. The
+    // actual parsing code will be written in the next PR, where we can validate
+    // the parsing result.
+    StatusOr<BloomFilter> maybe_bloom_filter =
+        BloomFilter::Create(std::vector<uint8_t>{}, 0, 0);
+    if (maybe_bloom_filter.ok()) {
```
[the remark this answers is not in the document]
Comment (milaGGL) [reply Firestore/core/src/remote/serializer.cc:1450]: Since the spec tests in this PR doesn't actually has Bloom filter, and not doing the parsing/serializing, the bloom filter creation is not tested. This part is covered in the next PR.

File: Firestore/core/src/remote/bloom_filter.cc [hunk 13]
```diff
@@ -115,8 +115,30 @@ bool BloomFilter::MightContain(absl::string_view value) const {
 
 bool operator==(const BloomFilter& lhs, const BloomFilter& rhs) {
   return lhs.bit_count() == rhs.bit_count() &&
-         lhs.hash_count() == rhs.hash_count() &&
-         equal(lhs.bitmap().begin(), lhs.bitmap().end(), rhs.bitmap().begin());
+         lhs.hash_count() == rhs.hash_count() && CompareBits(lhs, rhs);
+}
+
+bool operator!=(const BloomFilter& lhs, const BloomFilter& rhs) {
```
Comment (dconeybe) [inline Firestore/core/src/remote/bloom_filter.cc:121]: Put operator!= into the .h file, and declare it `inline`, like this:

https://github.com/firebase/firebase-ios-sdk/blob/b2383a9d3ca37b6a9e99c5fd9b75abdcf5a497c8/Firestore/core/src/model/transform_operation.h#L226-L229

File: Firestore/core/src/remote/bloom_filter.cc [hunk 14]
```diff
@@ -115,8 +115,30 @@ bool BloomFilter::MightContain(absl::string_view value) const {
...
+  return !(lhs == rhs);
+}
+
+bool CompareBits(const BloomFilter& lhs, const BloomFilter& rhs) {
```
Comment (dconeybe) [inline Firestore/core/src/remote/bloom_filter.cc:125]: Consider simplifying your CompareBits() function to do this:

1. Compare the bit_counts(), returning false if they are different.
2. Compare each _byte_ (each element) of the vectors _except_ for the last byte, returning false if any are different.
3. Compare the _last_ byte of the vectors, zeroing out the padding bits, returning false if they are different.
4. Return true

What do you think?

File: Firestore/core/src/remote/bloom_filter.h [hunk 15]
```diff
@@ -113,6 +113,10 @@ class BloomFilter final {
 
 bool operator==(const BloomFilter& lhs, const BloomFilter& rhs);
 
+bool operator!=(const BloomFilter& lhs, const BloomFilter& rhs);
+
+bool CompareBits(const BloomFilter& lhs, const BloomFilter& rhs);
```
Comment (dconeybe) [inline Firestore/core/src/remote/bloom_filter.h:118]: Please remove CompareBits() from the header file since it's just an implementation detail. In fact, it should be put into an anonymous namespace in the .cc file to scope it to the file.

File: Firestore/core/src/remote/bloom_filter.cc [hunk 16]
```diff
@@ -115,30 +144,7 @@ bool BloomFilter::MightContain(absl::string_view value) const {
...
-    const uint8_t& rhs_byte = rhs.bitmap()[i / 8];
-    const bool bit1 = (lhs_byte & (static_cast<uint8_t>(0x01) << offset));
-    const bool bit2 = (rhs_byte & (static_cast<uint8_t>(0x01) << offset));
-    if (bit1 != bit2) {
-      return false;
-    }
-  }
-
-  return true;
+         lhs.hash_count() == rhs.hash_count() && HasSameBits(lhs, rhs);
```
Comment (dconeybe) [inline Firestore/core/src/remote/bloom_filter.cc:147]: nit: Remove the check for `lhs.bit_count() == rhs.bit_count()` because that is already taken into account in `HasSameBits()`.

**Review by dconeybe — CHANGES_REQUESTED** [review]
Just one tiny comment!

File: Firestore/Example/Tests/SpecTests/FSTSpecTests.mm [hunk 17]
```diff
@@ -325,6 +329,14 @@ - (SnapshotVersion)parseVersion:(NSNumber *_Nullable)version {
   return Version(version.longLongValue);
 }
 
+- (absl::optional<BloomFilter>)parseBloomFilter:(NSDictionary *_Nullable)bloomFilterProto {
+  // TODO(Mila): None of the ported spec tests actually has the bloom filter json string, so hard
+  // code a null value for now. Actual parsing code will be written in the next PR, where we can
+  // validate the parsing result.
+  return absl::nullopt;
+  ;
```
Comment (dconeybe) [inline Firestore/Example/Tests/SpecTests/FSTSpecTests.mm:337]: nit: remove this unnecessary semicolon

File: Firestore/Example/Tests/SpecTests/FSTSpecTests.mm [hunk 18]
```diff
@@ -325,6 +329,14 @@ - (SnapshotVersion)parseVersion:(NSNumber *_Nullable)version {
   return Version(version.longLongValue);
 }
 
+- (absl::optional<BloomFilter>)parseBloomFilter:(NSDictionary *_Nullable)bloomFilterProto {
```
Comment (dconeybe) [inline Firestore/Example/Tests/SpecTests/FSTSpecTests.mm:332]: nit: Make `parseBloomFilter()` a free function defined in the anonymous namespace near the top of this file. As far as I can tell, it likely doesn't need to be a member function in the class; however, if I'm wrong then feel free to leave it as a member function.

File: Firestore/core/src/remote/bloom_filter.cc (hunk 16 above)
Comment (milaGGL) [reply Firestore/core/src/remote/bloom_filter.cc:147]: but, bloom filter has 3 attributes, even tho bit_count is checked inside HasSameBits, it looks nice to check all attributes in == operator.

File: Firestore/core/src/remote/serializer.cc [hunk 19]
```diff
@@ -1429,9 +1430,30 @@ std::unique_ptr<WatchChange> Serializer::DecodeDocumentRemove(
 
 std::unique_ptr<WatchChange> Serializer::DecodeExistenceFilterWatchChange(
     ReadContext*, const google_firestore_v1_ExistenceFilter& filter) const {
-  ExistenceFilter existence_filter{filter.count};
-  return absl::make_unique<ExistenceFilterWatchChange>(existence_filter,
-                                                       filter.target_id);
+  ExistenceFilter existence_filter = DecodeExistenceFilter(filter);
+  return absl::make_unique<ExistenceFilterWatchChange>(
+      std::move(existence_filter), filter.target_id);
```
Comment (dconeybe) [inline Firestore/core/src/remote/serializer.cc:1435]: Consider replacing `std::move(existence_filter)` with just `DecodeExistenceFilter(filter)`, avoiding the `existence_filter` local variable.

File: Firestore/core/src/remote/serializer.cc [hunk 20]
```diff
@@ -1429,9 +1430,30 @@ std::unique_ptr<WatchChange> Serializer::DecodeDocumentRemove(
...
+}
+
+ExistenceFilter Serializer::DecodeExistenceFilter(
+    const google_firestore_v1_ExistenceFilter& filter) const {
+  // Create bloom filter if there is an unchanged_names present in the filter
+  // and inputs are valid, otherwise keep it null.
+  absl::optional<BloomFilter> bloom_filter;
+  if (filter.has_unchanged_names && filter.unchanged_names.has_bits) {
```
Comment (dconeybe) [inline Firestore/core/src/remote/serializer.cc:1443]: Remove the check for `filter.unchanged_names.has_bits` here. Later on, when you actually implement this function, treat `has_bits==false` as if an empty bits with padding=0 were specified.

**Review by dconeybe — CHANGES_REQUESTED** [review]
Ok, a few more comments!

File: Firestore/core/src/remote/bloom_filter.cc (hunk 16 above)
Comment (dconeybe) [reply Firestore/core/src/remote/bloom_filter.cc:147]: All right. If you feel strongly enough about this to write a comment then feel free to leave it as is :)

Comment (milaGGL) [reply Firestore/core/src/remote/bloom_filter.cc:147]: not that strong

**Review by dconeybe — APPROVED** [review]
